### ISSUE TITLE
[Feature]: Hardware wallet type field

### ISSUE THREAD

**Issue by leovct** [blocknative/web3-onboard#897]
### Is your request related to a problem?

The old library *bnc-onboard* provided a wallet type field ([link to the old doc](https://docs.blocknative.com/onboard-older-version#subscriptions-optional)) that is not available in the new *web3-onboard* library. I may have missed it but I don't see anything related to hardware wallets in the [new documentation](https://docs.blocknative.com/onboard/core#state).

For reference, here is the old wallet spec:
- provider - The JavaScript provider for interacting with the wallet
- name - The wallet display name
- instance - If the wallet type is 'sdk' then this is the initialized wallet instance
- **type - The wallet type 'hardware' | 'injected' | 'sdk'**
- connect - The function that initiates the wallet connection logic
- dashboard - Some SDK wallets allow for opening to wallet dashboard
- icons - [object] Image strings for the wallet icon { svg, src, srcset }

### Feature Description

Add a `type` field to the `WalletState` type. The possible values should be the same as those of the previous version, namely `hardware`, `injected` or `sdk`.

```js
type WalletState = {
  label: string
  icon: string
  provider: EIP1193Provider
  accounts: Account[]
  chains: ConnectedChain[]
  instance?: unknown
  type: WalletType // This is the new field.
}

type WalletType = 'hardware' | 'injected' | 'sdk';
```

**Comment by aaronbarnardsound** [comment]
Hey @leovct we have tried to standardize all of the wallet providers so that the `type` field was no longer needed. Can I ask what you are using the `type` field on the wallets for and how it is impacting your app not having it there?

**Comment by leovct** [comment]
Thanks for your reply!

In our application, we treated hardware wallets differently from the rest of the wallets by allowing the user to switch networks if the wallet was of hardware type. For the other wallets, the _bnc-onboard_ library would manage the network switch natively.

In fact, I'm migrating our authentication component from _bnc-onboard_ to _web3-onboard_ and the procedure is not detailed at all in the documentation. So as soon as I saw that the option we were using before was not present in the new version, I opened a ticket. In the end you're right, the library is much more standardized now and _web3-onboard_ already handles this feature natively. All good! :smile:

**Comment by whatsyourgithub** [comment]
@aaronbarnardsound We have an issue with a standard interface and no type. You can send transactions to wallets like MetaMask without fields like nonce and gas, because they fill those in before signing the tx. But leaving them out results in an error from the Trezor, Ledger, and KeepKey modules, because they don't provide them. Also, the typescript definitions are the same across all wallet types, which says those fields are optional, despite causing an error every single time on hardware wallets. Because there is no way to detect wallet type, it is difficult to tell when our app needs to handle nonce, gas, etc or when we should allow the wallet to handle them.

**Comment by aaronbarnardsound** [comment]
Hey @whatsyourgithub yeah that is a good point. Are you using the `eth_sendTransaction` method directly then on all wallets? Libraries like ethers populate those transaction fields for you, but we could always add that functionality to hardware wallets if you think that would be useful? Otherwise at the least we should update the types to be correct for sure.

**Comment by aaronbarnardsound** [comment]
> In fact, I'm migrating our authentication component from bnc-onboard to web3-onboard and the procedure is not detailed at all in the documentation

@leovct do you think that we need some specific documentation for migrating from v1 to v2? If you have any notes or thoughts from your migration, that would be super helpful. It is hard when you are deep into implementing the code to see where the documentation is lacking :)

**Comment by leovct** [comment]
> > In fact, I'm migrating our authentication component from bnc-onboard to web3-onboard and the procedure is not detailed at all in the documentation
> 
> @leovct do you think that we need some specific documentation for migrating from v1 to v2? If you have any notes or thoughts from your migration, that would be super helpful. It is hard when you are deep into implementing the code to see where the documentation is lacking :)

@aaronbarnardsound Absolutely! A migration guide would be really useful I think. Do you want me to open a new ticket for this?
At the moment, I'm not done yet with the migration but once it's working, i'd happy to help you on this task.

**Comment by whatsyourgithub** [comment]
@aaronbarnardsound We are wrapping the provider in an ethers Web3Provider, which does not provide all missing fields

**Comment by aaronbarnardsound** [comment]
> Absolutely! A migration guide would be really useful I think. Do you want me to open a new ticket for this?

Yes that would be great and once you have migrated, a draft from your perspective would be really helpful!

**Comment by aaronbarnardsound** [comment]
> @aaronbarnardsound We are wrapping the provider in an ethers Web3Provider, which does not provide all missing fields

Ah yes. I thought the ethers provider was supposed to call `signer.populateTransaction` automatically, but I have noticed the same behavior. I think we will go ahead and add extra functionality to the hardware wallets to populate those missing fields and we are looking in to using our [Gas Platform](https://www.blocknative.com/gas-platform) for the gas estimations as well!

**Comment by rbruels** [comment]
> Hey @whatsyourgithub yeah that is a good point. Are you using the `eth_sendTransaction` method directly then on all wallets? Libraries like ethers populate those transaction fields for you, but we could always add that functionality to hardware wallets if you think that would be useful? Otherwise at the least we should update the types to be correct for sure.

Just wanted to chime in here — yes, this would be useful!

**Comment by aaronbarnardsound** [comment]
Closed via #962 
Rather than adding a label, we have made it so that hardware wallets will automatically populate missing fields for convenience.
